code/modules/power/fuel_assembly/fuel_compressor.dm

```diff
@@ -0,0 +1,146 @@
+	if(!ui)
+		ui = new(user, src, ui_key, "fuel_compressor.tmpl", name, 500, 600)
+		ui.set_initial_data(data)
+		ui.open()
+
+/obj/machinery/fuel_compressor/OnTopic(user, href_list, state)
+	. = ..()
+	if(href_list["eject_material"])
+		. = eject_material(href_list["eject_material"])
+	if(href_list["make_rod"])
+		. = make_rod()
+	if(href_list["change_makeup"])
+		. = change_makeup(href_list["change_makeup"], user)
+	if(href_list["clear_makeup"])
+		rod_makeup.Cut()
+		return TOPIC_REFRESH
+
+/obj/machinery/fuel_compressor/proc/eject_material(var/mat_name)
+	var/mat_type
+	for(var/mat_p in stored_material)
+		var/decl/material/mat = decls_repository.get_decl(mat_p)
+		if(mat.name == mat_name)
+			mat_type = mat.type
+			break
+
+	if(!mat_type)
+		return TOPIC_HANDLED
+
+	var/decl/material/mat = decls_repository.get_decl(mat_type)
+	if(mat && stored_material[mat_type] >= SHEET_MATERIAL_AMOUNT)
+		var/sheet_count = Floor(stored_material[mat_type]/SHEET_MATERIAL_AMOUNT)
+		stored_material[mat_type] -= sheet_count * SHEET_MATERIAL_AMOUNT
+		SSmaterials.create_object(mat_type, get_turf(src), sheet_count)
+		if(isnull(stored_material[mat_type]))
+			stored_material -= mat_type
+	else if(!isnull(stored_material[mat_type]))
+		stored_material -= mat_type
+	
+	return TOPIC_REFRESH
+
+/obj/machinery/fuel_compressor/proc/make_rod()
+	var/total_matter = 0
+	for(var/mat_p in rod_makeup)
+		total_matter += rod_makeup[mat_p]
+		if(total_matter > MAX_ROD_MATERIAL)
+			visible_message(SPAN_WARNING("\The [src] flashes an 'Over max material' error!"))
+			return TOPIC_HANDLED
+		if(rod_makeup[mat_p] > stored_material[mat_p])
+			visible_message(SPAN_WARNING("\The [src] flashes an 'Insufficient Materials' error!"))
+			return TOPIC_HANDLED
+	
+	for(var/mat_p in rod_makeup)
+		stored_material[mat_p] -= rod_makeup[mat_p]
+		if(stored_material[mat_p] == 0)
+			stored_material -= mat_p
+	
+	visible_message(SPAN_NOTICE("\The [src] compresses the material into a new fuel assembly."))
+	new /obj/item/fuel_assembly(get_turf(src), rod_makeup)
+	return TOPIC_REFRESH
+
+/obj/machinery/fuel_compressor/proc/change_makeup(var/mat_name, var/mob/user)
+	var/mat_type
+	for(var/mat_p in stored_material)
+		var/decl/material/mat = decls_repository.get_decl(mat_p)
+		if(mat.name == mat_name)
+			mat_type = mat.type
+			break
+
+	if(!mat_type)
+		return TOPIC_HANDLED
+	var/amt = input(user, "Enter the amount of this material per rod (Max [MAX_ROD_MATERIAL]):", "Fuel Rod Makeup", rod_makeup[mat_type]) as null|num
+	if(!CanInteract(user, DefaultTopicState()))
+		return TOPIC_HANDLED
+	amt = round(Clamp(amt, 0, MAX_ROD_MATERIAL))
+	if(!amt)
+		rod_makeup -= mat_type
+		return TOPIC_REFRESH			
+	rod_makeup[mat_type] = amt
+	return TOPIC_REFRESH
+
+/obj/machinery/fuel_compressor/receive_mouse_drop(var/atom/movable/dropping, var/mob/user)
+	if(user.incapacitated() || !user.Adjacent(src))
+		return
+	return add_material(dropping, user)
```

Comment:
`return !add_material(dropping, user)` - truthy return value incidates stuff further up the chain can keep working with `dropping`.